src/ApiCore/ResourceTemplate/Segment.php

```diff
@@ -0,0 +1,159 @@
+<?php
+/*
+ * Copyright 2018 Google LLC
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+namespace Google\ApiCore\ResourceTemplate;
+
+use Google\ApiCore\ValidationException;
+
+/**
+ * Represents a segment in a resource template. This is used internally by RelativeResourceTemplate,
+ * but is not intended for public use and may change without notice.
+ * @internal
+ */
+class Segment
+{
+    const LITERAL_SEGMENT = 0;
+    const WILDCARD_SEGMENT = 1;
+    const DOUBLE_WILDCARD_SEGMENT = 2;
+    const VARIABLE_SEGMENT = 3;
+
+    /** @var int */
+    private $segmentType;
+
+    /** @var string|null */
+    private $key;
+
+    /** @var string|null */
+    private $value;
+
+    /** @var RelativeResourceTemplate|null */
+    private $template;
+
+    public function __construct($segmentType, $key, $value, RelativeResourceTemplate $template = null)
```

Comment:
Since this is internal.. no big deal - but just a note that it may be handy to add a docblock since this is public

Reply:
Fixed in https://github.com/googleapis/gax-php/pull/209